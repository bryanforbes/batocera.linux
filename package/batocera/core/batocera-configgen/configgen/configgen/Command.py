from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
