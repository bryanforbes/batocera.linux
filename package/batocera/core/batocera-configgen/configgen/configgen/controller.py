from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
