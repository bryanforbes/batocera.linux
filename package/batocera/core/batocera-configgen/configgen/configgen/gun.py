from __future__ import annotations

import logging
import re
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Final, cast

import evdev

_input_re = re.compile(r"^/dev/input/event([0-9]*)$")
_logger = logging.getLogger(__name__)

_mouse_code_to_button: Final = {
    evdev.ecodes.BTN_LEFT: 'left',
    evdev.ecodes.BTN_RIGHT: 'right',
    evdev.ecodes.BTN_MIDDLE: 'middle',
    evdev.ecodes.BTN_1: '1',
    evdev.ecodes.BTN_2: '2',
    evdev.ecodes.BTN_3: '3',
    evdev.ecodes.BTN_4: '4',
    evdev.ecodes.BTN_5: '5',
    evdev.ecodes.BTN_6: '6',
    evdev.ecodes.BTN_7: '7',
    evdev.ecodes.BTN_8: '8',
}

_mouse_button_to_code: Final = {button: code for code, button in _mouse_code_to_button.items()}

def _get_mouse_buttons(device: evdev.InputDevice) -> list[str]:
    caps_keys = device.capabilities()[evdev.ecodes.EV_KEY]
    mouse_caps = set(caps_keys) & set(_mouse_code_to_button.keys())

    return [button for cap, button in _mouse_code_to_button.values() if cap in mouse_caps]


@dataclass(slots=True, kw_only=True)
class Gun:
    node: str
    mouse_index: int
    need_cross: bool
    need_borders: bool
    name: str
    buttons: list[str]

    @classmethod
    def get_all(cls) -> GunDict:
        import pyudev

        guns: GunDict = {}
        context = pyudev.Context()

        mouses = {
            int(match.group(1)): mouse
            for mouse in context.list_devices(subsystem='input')
            if mouse.device_node is not None and
            (match := _input_re.match(mouse.device_node)) is not None and
            mouse.properties.get("ID_INPUT_MOUSE") == '1'
        }

        gun_index = 0
        for mouse_index, (_, mouse) in enumerate(sorted(mouses.items(), key=lambda item: item[0])):
            _logger.info("found mouse %s at %s with mouse_index=%s", mouse_index, mouse.device_node, mouse_index)
            if "ID_INPUT_GUN" not in mouse.properties or mouse.properties["ID_INPUT_GUN"] != "1":
                continue

            device = evdev.InputDevice(cast(str, mouse.device_node))

            guns[gun_index] = cls(
                node=cast(str, mouse.device_node),
                mouse_index=mouse_index,
                # retroarch uses mouse indexes into configuration files using ID_INPUT_MOUSE (TOUCHPAD are listed after mouses)
                need_cross=mouse.properties.get("ID_INPUT_GUN_NEED_CROSS") == '1',
                need_borders=mouse.properties.get("ID_INPUT_GUN_NEED_BORDERS") == '1',
                name=device.name,
                buttons=_get_mouse_buttons(device)
            )

            gun_index += 1

        if not guns:
            _logger.info("no guns found")

        return guns


def gun_button_to_code(button: str) -> int | None:
    return _mouse_button_to_code.get(button)


def guns_need_crosses(guns: GunMapping) -> bool:
    # no gun, enable the cross for joysticks, mouses...
    if not guns:
        return True

    return any(gun.need_cross for gun in guns.values())

def guns_borders_size_name(guns: GunMapping, config: Mapping[str, object]) -> str | None:
    borders_size: str = cast(str, config.get("controllers.guns.borderssize", "medium"))

    # overriden by specific options
    borders_mode = "normal"
    if (config_borders_mode := cast(str, config.get("controllers.guns.bordersmode", "auto"))) != "auto":
        borders_mode = config_borders_mode
    if (config_borders_mode := cast(str, config.get("bordersmode", "auto"))) != "auto":
        borders_mode = config_borders_mode

    # others are gameonly and normal
    if borders_mode == "hidden":
        return None

    if borders_mode == "force":
        return borders_size

    for gun in guns.values():
        if gun.need_borders:
            return borders_size

    return None

# returns None to follow the bezel overlay size by default
def guns_border_ratio_type(guns: GunMapping, config: dict[str, str]) -> str | None:
    if "controllers.guns.bordersratio" in config:
        return config["controllers.guns.bordersratio"] # "4:3"
    return None


type GunMapping = Mapping[int, Gun]
type GunDict = dict[int, Gun]
