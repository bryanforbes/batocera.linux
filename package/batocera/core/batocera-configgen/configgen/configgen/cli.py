from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
