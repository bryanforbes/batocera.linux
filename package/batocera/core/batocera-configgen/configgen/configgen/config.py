from __future__ import annotations

import logging
import xml.etree.ElementTree as ET
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Self, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path

    from .cli import Options
    from .settings.unixSettings import UnixSettings

_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')

_logger = logging.getLogger(__name__)


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: Mapping[str, Any]) -> None:
    """Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if key in destination and isinstance(destination[key], dict) and isinstance(value, Mapping):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


def _load_defaults(system: str, default_file: Path, default_arch_file: Path, /) -> dict[str, Any]:
    with default_file.open() as f:
        defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

    with default_arch_file.open() as f:
        arch_defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader) or {}

    config: dict[str, Any] = defaults.get('default', {})

    if 'default' in arch_defaults:
        _dict_merge(config, arch_defaults['default'])

    if system in defaults:
        _dict_merge(config, defaults[system])

    if system in arch_defaults:
        _dict_merge(config, arch_defaults[system])

    return config


@dataclass(slots=True)
class Config:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    data: dict[str, Any]

    def __contains__(self, x: object, /) -> bool:
        return x in self.data

    @overload
    def get(self, key: str, default: MissingType = ..., /) -> Any | MissingType: ...

    @overload
    def get(self, key: str, default: _T, /) -> Any | _T: ...

    def get(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.data.get(key, default)

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool: ...

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV: ...

    def get_bool(
        self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None
    ) -> bool | _TV | _FV:
        value = self.get(key)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES

        return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_str(self, key: str, default: MissingType = ..., /) -> str | MissingType: ...

    @overload
    def get_str(self, key: str, default: str, /) -> str: ...

    def get_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_int(self, key: str, default: MissingType = ..., /) -> int | MissingType: ...

    @overload
    def get_int(self, key: str, default: int, /) -> int: ...

    def get_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_float(self, key: str, default: MissingType = ..., /) -> float | MissingType: ...

    @overload
    def get_float(self, key: str, default: float, /) -> float: ...

    def get_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return float(value)

    def items(self, /, *, starts_with: str | None = None) -> Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.data.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.data.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


def _get_rom_settings_name(rom: Path, /) -> str:
    # sanitize rule by EmulationStation
    # see FileData::getConfigurationName() on batocera-emulationstation
    rom_settings_name = rom.name.replace('=', '').replace('#', '')

    _logger.info('game settings name: %s', rom_settings_name)

    return rom_settings_name


@dataclass
class SystemConfig(Config):
    @property
    def core(self) -> str:
        return self.data['core']

    @property
    def core_forced(self) -> str:
        return self.data['core-forced']

    @property
    def emulator(self) -> str:
        return self.data['emulator']

    @property
    def emulator_forced(self) -> str:
        return self.data['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.data['uimode']

    @property
    def show_fps(self) -> bool:
        return self.data['showFPS']

    @property
    def video_mode(self) -> str:
        return self.data.get('videomode', 'default')

    @property
    def use_guns(self) -> bool:
        return self.get_bool('use_guns')

    @property
    def use_wheels(self) -> bool:
        return self.get_bool('use_wheels')

    @classmethod
    def load(cls, options: Options, settings: UnixSettings, /) -> Self:
        defaults = _load_defaults(
            options['system'], DEFAULTS_DIR / 'configgen-defaults.yml', DEFAULTS_DIR / 'configgen-defaults-arch.yml'
        )
        data: dict[str, Any] = {
            'emulator': defaults['emulator'],
            'core': defaults['core'],
        }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(data, defaults['options'])

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        data.update(settings.get_all_iter('display', keep_name=True, keep_defaults=True))
        data.update(settings.get_all_iter('controllers', keep_name=True))

        global_settings = settings.get_all('global')
        system_settings = settings.get_all(options['system'])
        folder_settings = settings.get_all(f'{options["system"]}.folder["{rom.parent}"]')
        game_settings = settings.get_all(f'{options["system"]}["{rom_settings_name}"]')

        data.update(global_settings)
        data.update(system_settings)
        data.update(folder_settings)
        data.update(game_settings)

        try:
            es_config = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = es_config.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib['value'] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            data['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = es_config.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib['value'] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            data['uimode'] = uimode_value
        except Exception:
            data['showFPS'] = False
            data['uimode'] = 'Full'

        data['emulator-forced'] = (
            'emulator' in global_settings
            or 'emulator' in system_settings
            or 'emulator' in game_settings
            or options['emulator'] is not None
        )
        data['core-forced'] = (
            'core' in global_settings
            or 'core' in system_settings
            or 'core' in game_settings
            or options['core'] is not None
        )

        if options['emulator'] is not None:
            data['emulator'] = options['emulator']

        if options['core'] is not None:
            data['core'] = options['core']

        if 'use_guns' not in data and options['lightgun']:
            data['use_guns'] = True

        if 'use_wheels' not in data and options['wheel']:
            data['use_wheels'] = True

        # network options
        if options['netplaymode'] is not None:
            data['netplay.mode'] = options['netplaymode']

        if options['netplaypass'] is not None:
            data['netplay.password'] = options['netplaypass']

        if options['netplayip'] is not None:
            data['netplay.server.ip'] = options['netplayip']

        if options['netplayport'] is not None:
            data['netplay.server.port'] = options['netplayport']

        if options['netplaysession'] is not None:
            data['netplay.server.session'] = options['netplaysession']

        # autosave arguments
        if options['state_slot'] is not None:
            data['state_slot'] = options['state_slot']

        if options['autosave'] is not None:
            data['autosave'] = options['autosave']

        if options['state_filename'] is not None:
            data['state_filename'] = options['state_filename']

        return cls(data)


@dataclass
class RenderConfig(Config):
    @classmethod
    def load(cls, options: Options, settings: UnixSettings, shader_set: str | MissingType, /) -> Self:
        data: dict[str, Any] = {}

        if shader_set is not cls.MISSING:
            if shader_set == 'none':
                rendering_defaults = BATOCERA_SHADERS / 'configs' / 'rendering-defaults.yml'
            else:
                rendering_defaults = USER_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'
                if not rendering_defaults.exists():
                    rendering_defaults = BATOCERA_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'

            data = _load_defaults(
                options['system'], rendering_defaults, rendering_defaults.with_name('rendering-defaults-arch.yml')
            )

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        # es only allow to update system and game settings in fact for the moment

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        data.update(settings.get_all_iter(f'{options["system"]}-renderer'))
        data.update(settings.get_all_iter(f'{options["system"]}["{rom_settings_name}"]-renderer'))

        return cls(data)
