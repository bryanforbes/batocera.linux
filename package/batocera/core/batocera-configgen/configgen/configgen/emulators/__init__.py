from __future__ import annotations

from contextlib import nullcontext
from importlib import import_module
from typing import TYPE_CHECKING, Final

from ..batoceraPaths import BATOCERA_CONF
from ..config import RenderConfig, SystemConfig
from ..controller import Controller
from ..controllersConfig import getGamesMetaData
from ..gun import Gun, precalibrate_guns
from ..settings.unixSettings import UnixSettings
from ..utils.squashfs import squashfs_rom
from ..utils.wheels import configure_wheels
from ._base import Emulator as Emulator  # noqa: TCH001

if TYPE_CHECKING:
    from ..cli import Options

_GENERATOR_MAP: Final[dict[str, str]] = {}


def get_emulator_cls(emulator: str, /) -> type[Emulator]:
    module_path = _GENERATOR_MAP.get(emulator, emulator.replace('-', '_'))

    try:
        module = import_module(f'..{module_path}', package=__name__)
        emulator_cls: type[Emulator] = module.Emulator
    except ImportError as e:
        if e.name is not None and e.name.startswith(__name__.split('.')[0]):
            raise Exception(f'no emulator found for {emulator!r}') from e

        raise
    except AttributeError as e:
        raise Exception(f'no emulator found for {emulator!r}') from e

    return emulator_cls


def run_emulator(options: Options, /) -> int:
    exit_code = -1
    original_rom = options['rom']

    if original_rom.suffix == '.squashfs':
        rom_context_manager = squashfs_rom(original_rom)
    else:
        rom_context_manager = nullcontext(original_rom)

    with rom_context_manager as rom:
        system = options['system']

        settings = UnixSettings(BATOCERA_CONF)
        system_config = SystemConfig.load(options, settings)
        render_config = RenderConfig.load(options, settings, system_config.get('shaderset'))

        metadata = getGamesMetaData(system, original_rom)

        controllers = Controller.load_for_player_options(options['players'])

        if system_config.use_guns:
            guns = Gun.get_all()
            precalibrate_guns(system, system_config.emulator, system_config.core, original_rom)
        else:
            guns = {}

        with configure_wheels(controllers, system, system_config, metadata) as (controllers, wheels):
            emulator_cls = get_emulator_cls(system_config.emulator)
            emulator = emulator_cls(system, system_config, render_config, metadata, controllers, guns, wheels)
            exit_code = emulator.run(rom, original_rom, options['gameinfoxml'])

    return exit_code
