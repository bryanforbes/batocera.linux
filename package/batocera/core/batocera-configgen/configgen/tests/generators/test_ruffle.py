from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import ROMS
from configgen.config import SystemConfig
from configgen.generators.ruffle.ruffleGenerator import RuffleGenerator
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from pytest_mock import MockerFixture
    from syrupy.assertion import SnapshotAssertion


class TestRuffleGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[RuffleGenerator]:
        return RuffleGenerator

    def test_get_mouse_mode(self, generator: RuffleGenerator) -> None:  # pyright: ignore
        assert generator.getMouseMode(SystemConfig({}), Path())

    def test_generate(
        self,
        generator: RuffleGenerator,
        mocker: MockerFixture,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mocker.Mock(),
                ROMS / 'flash' / 'rom.swf',
                [],
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
