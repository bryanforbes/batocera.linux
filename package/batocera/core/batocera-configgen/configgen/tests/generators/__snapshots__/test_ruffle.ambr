# serializer version: 1
# name: TestRuffleGenerator.test_generate
  tuple(
    list([
      'ruffle',
      '/userdata/roms/flash/rom.swf',
    ]),
    dict({
    }),
  )
# ---
# name: TestRuffleGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
    }),
    'name': 'ruffle',
  })
# ---
